contracts/mesh-provider/src/ibc.rs

```diff
@@ -213,33 +223,110 @@ pub fn ack_list_validators(
     for val in res.validators {
         VALIDATORS.save(deps.storage, &val, &Validator::new())?;
     }
+
+    // reset retry counter
+    let mut config = CONFIG.load(deps.storage)?;
+    config.list_validators_retries_remaining = LIST_VALIDATORS_RETRIES;
+    CONFIG.save(deps.storage, &config)?;
+
     Ok(IbcBasicResponse::new())
 }
 
-pub fn fail_list_validators(_deps: DepsMut) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: send another ListValidators message
-    unimplemented!();
+pub fn fail_list_validators(deps: DepsMut, env: Env) -> Result<IbcBasicResponse, ContractError> {
+    // check if we should retry
+    let cfg = CONFIG.load(deps.storage)?;
+    let channel_id = CHANNEL.load(deps.storage)?;
+    if cfg.list_validators_retries_remaining <= 0 {
+        return Err(ContractError::NoResponse("list_validators".to_string(), channel_id))
+    }
+    let mut cfg_after = cfg.clone();
+    cfg_after.list_validators_retries_remaining -= 1;
+    CONFIG.save(deps.storage, &cfg_after)?;
+
+    // retry
+    let packet = ProviderMsg::ListValidators {};
+    let msg = IbcMsg::SendPacket {
+        channel_id: channel_id,
+        data: to_binary(&packet)?,
+        timeout: build_timeout(&env),
+    };
+    Ok(IbcBasicResponse::new().add_message(msg))
 }
 
 pub fn fail_stake(
-    _deps: DepsMut,
+    deps: DepsMut,
     // _staker is the staker Addr, not used by consumer
-    _staker: String,
-    _validator: String,
-    _amount: Uint128,
+    staker: String,
+    validator: String,
+    amount: Uint128,
 ) -> Result<IbcBasicResponse, ContractError> {
-    // TODO: release the bonded stake, adjust numer
-    unimplemented!();
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
+    }
+
+    let mut val = VALIDATORS
+        .may_load(deps.storage, &validator)?
+        .ok_or_else(|| ContractError::UnknownValidator(validator.clone()))?;
+    let mut stake = STAKED
+        .may_load(deps.storage, (&staker, &validator))?
+        .unwrap_or_default();
+    stake.unstake_validator(&mut val, amount)?;
+    STAKED.save(deps.storage, (&staker, &validator), &stake)?;
+    VALIDATORS.save(deps.storage, &validator, &val)?;
+    Ok(IbcBasicResponse::new())
+}
+
+pub fn ack_unstake(
+    deps: DepsMut,
+    env: Env,
+    staker: String,
+    _validator: String,
+    amount: Uint128
+) -> Result<IbcBasicResponse , ContractError> {
+    let staker = deps.api.addr_validate(&staker)?;
+
+    if amount.is_zero() {
+        return Err(ContractError::ZeroAmount);
+    }
+
+    // create a future claim on number of shares (so we can adjust for later slashing)
+    let cfg = CONFIG.load(deps.storage)?;
+    let ready = env.block.time.plus_seconds(cfg.unbonding_period);
+    CLAIMS.create_claim(
+        deps.storage,
+        &staker,
+        amount,
+        Expiration::AtTime(ready),
+    )?;
+
+    // TODO: How should the user be notified of the failure?
```

Comment:
```suggestion
```

Reply:
Let's just add some event to the Responses here (and make them successes).

How about defining some standard event type to hold all errors (it is just arbitrary keys, but spec'ing it in the readme helps). Then something like:

```rust
let evt = Event::new("mesh_ibc_error")
  .add_attribute("action", "unstake")
  .add_attribute("error", err);
```

(Where `err` is any `ContractError` created earlier in the handling)